pontoon/contributors/utils.py

```diff
@@ -336,21 +328,170 @@ def get_contributions(user, contribution_type=None):
         )
     )
 
-    action_map = {
+    return {
         "user_translations": user_translations,
         "user_reviews": user_reviews,
         "peer_reviews": peer_reviews,
         "all_user_contributions": all_user_contributions,
         "all_contributions": all_contributions,
     }
 
-    if contribution_type not in action_map.keys():
+
+def get_contribution_graph_data(user, contribution_type=None):
+    """
+    Get data required to render the Contribution graph on the Profile page
+    """
+    contribution_period = Q(created_at__gte=timezone.now() - relativedelta(days=365))
+    contributions_map = get_contributions_map(user, contribution_period)
+
+    if contribution_type not in contributions_map.keys():
         contribution_type = "all_user_contributions"
 
-    contributions = get_daily_action_counts(action_map[contribution_type])
-    total = sum(contributions.values())
+    contributions_qs = contributions_map[contribution_type]
+    contributions_data = {
+        convert_to_unix_time(item["timestamp"]): item["count"]
+        for item in (
+            contributions_qs.annotate(timestamp=TruncDay("created_at"))
+            .values("timestamp")
+            .annotate(count=Count("id"))
+            .values("timestamp", "count")
+        )
+    }
+
+    total = sum(contributions_data.values())
 
     return (
-        contributions,
+        contributions_data,
         f"{ intcomma(total) } contribution{ pluralize(total) } in the last year",
     )
+
+
+def get_project_locale_contribution_counts(contributions_qs):
+    counts = {}
+
+    for item in (
+        contributions_qs.annotate(
+            project_name=F("translation__entity__resource__project__name"),
+            project_slug=F("translation__entity__resource__project__slug"),
+            locale_name=F("translation__locale__name"),
+            locale_code=F("translation__locale__code"),
+        )
+        .values("project_name", "project_slug", "locale_name", "locale_code")
+        .annotate(count=Count("id"))
```

Comment:
```suggestion
        contributions_qs.annotate(
            project_name=F("translation__entity__resource__project__name"),
            project_slug=F("translation__entity__resource__project__slug"),
            locale_name=F("translation__locale__name"),
            locale_code=F("translation__locale__code"),
            count=Count("id"),
        )
```

Would this also work? Not sure tbh, but as is the code looks awfully repetitive.

Reply:
Sadly no.

To quote the [docs](https://docs.djangoproject.com/en/4.1/topics/db/aggregation/#values):

> Ordinarily, annotations are generated on a per-object basis - an annotated QuerySet will return one result for each object in the original QuerySet. However, when a values() clause is used to constrain the columns that are returned in the result set, the method for evaluating annotations is slightly different. Instead of returning an annotated result for each result in the original QuerySet, the original results are **grouped according to the unique combinations of the fields specified in the values() clause**. An annotation is then provided for each unique group; the annotation is computed over all members of the group.